### PR TITLE
[Controller] Update CHIPController to use the Invoke API instead of r…

### PR DESCRIPTION
**Pull request by vivien-apple**
…elying on CHIPClusters

#### Problem

`CHIPDeviceController` and `CHIPDeviceCommissioner` stills uses APIs defines into `CHIPClusters.cpp`. As far as I know this is the last consumer of this code, which forces us to keeps `ChipClusters.cpp`, `ChipClientCallbacks.cpp`, some handler code into `IMClusterCommandHandler.cpp`. So removing it may allow to remove a lot of extra code.

#### Change overview
 * Replace IM calls to use the `Invoke` API

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -1891,21 +1646,20 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
         UpdateDevice(proxy->GetDeviceId());
     }
     break;
     case CommissioningStage::kSendComplete: {
-        ChipLogProgress(Controller, "Calling commissioning complete");
-        GeneralCommissioningCluster genCom;
-        genCom.Associate(proxy, 0);
-        genCom.CommissioningComplete(mSuccess.Cancel(), mFailure.Cancel());
+        GeneralCommissioning::Commands::CommissioningComplete::Type request;
+        StopSessionIf(SendCommand<NetworkCommissioningCluster>(proxy, request, OnSuccess<decltype(request)::ResponseType>));
```
Comment (msandstedt) [inline src/controller/CHIPDeviceController.cpp:1662]: I suspect CommissioningComplete and ArmFailSafe invocations will now fail.  As far as I know, accessories aren't returning correct responses for these.

See https://github.com/project-chip/connectedhomeip/issues/13327.

The legacy cluster code seemed to tolerate this.  I don't believe the new IM code will.

My solution in the #13294 feature branch was to simply tolerate non-success for these two invocaitons.

Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp:1662]: You "were" right until https://github.com/project-chip/connectedhomeip/commit/5cd7eee9810bc8c06a56ad3a2a5b1b57478a5ac0 ;)

Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp:1662]: Ah, that's great.  Thank you.

Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp:1662]: Yes, tried it myself.  The fix works great.  Thanks again.

**Review by msandstedt — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp [hunk 2]
```diff
@@ -602,13 +616,31 @@ ControllerDeviceInitParams DeviceController::GetControllerDeviceInitParams()
...
 
+#define TRAMPOLINE(Name, ArgType)                                                                                                  \
+    void DeviceCommissioner::On##Name(void * context, ArgType arg1)                                                                \
+    {                                                                                                                              \
+        ChipLogProgress(Controller, "Received %s callback", "Name");                                                               \
+        DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);                                            \
+        commissioner->On##Name(arg1);                                                                                              \
+    }
+
+#define TRAMPOLINE_2(Name, ArgType, ArgType2)                                                                                      \
```
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:627]: TRAMPOLINE_2 and 4 are called only once - using macros makes the code harder to follow. How about inlining them?

I am actually wondering if to just copy & paste these methods instead of usin TRAMPOLINE here to have less macro usage. Macros make code search hard (I will not be able to search for On... stuff) and are harder to update. Also wondering here if "ChipLogProgress" actually prints out the name or just says `Name` as it is unclear from the macro.

**Review by woody-apple — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp [hunk 3]
```diff
@@ -1001,7 +1033,7 @@ void DeviceCommissioner::OnSessionEstablishmentError(CHIP_ERROR err)
 
 void DeviceCommissioner::OnSessionEstablished()
 {
-    VerifyOrReturn(mDeviceBeingCommissioned != nullptr, OnSessionEstablishmentError(CHIP_ERROR_INVALID_DEVICE_DESCRIPTOR));
+    StopSessionIf(mDeviceBeingCommissioned == nullptr, CHIP_ERROR_INVALID_DEVICE_DESCRIPTOR);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1036]: My concern here is that it's very nonobvious that `StopSessionIf` is an early return.  The old naming of VerifyOrReturn at least made that clear...

Maybe: `StopSessionAndReturnIf`?

File: src/controller/CHIPDeviceController.cpp [hunk 4]
```diff
@@ -1034,166 +1061,73 @@ void DeviceCommissioner::OnSessionEstablished()
...
-void DeviceCommissioner::OnAttestationFailureResponse(void * context, uint8_t status)
-{
-    ChipLogProgress(Controller, "Device failed to receive the Attestation Information Response: 0x%02x", status);
-    DeviceCommissioner * commissioner = reinterpret_cast<DeviceCommissioner *>(context);
-    commissioner->mAttestationResponseCallback.Cancel();
-    commissioner->mOnAttestationFailureCallback.Cancel();
-    // TODO: Map error status to correct error code
-    commissioner->OnSessionEstablishmentError(CHIP_ERROR_INTERNAL);
+    OperationalCredentials::Commands::AttestationRequest::Type request;
+    request.attestationNonce = device->GetAttestationNonce();
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1104]: The code used to send `attestationNonce` here.  Why the change to `device->GetAttestationNonce()`?

File: src/controller/CHIPDeviceController.cpp [hunk 5]
```diff
@@ -1034,166 +1061,73 @@ void DeviceCommissioner::OnSessionEstablished()
...
-                         static_cast<uint16_t>(result));
-            // Go look at AttestationVerificationResult enum in src/credentials/DeviceAttestationVerifier.h to understand the
-            // errors.
-            SuccessOrExit(error = CHIP_ERROR_INTERNAL);
-        }
-    }
-
-    ChipLogProgress(Controller, "Successfully validated 'Attestation Information' command received from the device.");
+    StopSessionIf(mState != State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    StopSessionIf(mDeviceBeingCommissioned == nullptr, CHIP_ERROR_INCORRECT_STATE);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1122]: We didn't use to do these two checks inside `OnAttestationResponse`, right?  Why are we adding them?

Especially given that `ValidateAttestationInfo`, which we are about to call, will do them anyway.

File: src/controller/CHIPDeviceController.cpp [hunk 6]
```diff
@@ -1209,192 +1143,167 @@ CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(const ByteSpan & attestat
...
-    return CHIP_NO_ERROR;
 }
 
-void DeviceCommissioner::HandleAttestationResult(CHIP_ERROR err)
+TRAMPOLINE(DeviceAttestationInformationVerification, AttestationVerificationResult);
+void DeviceCommissioner::OnDeviceAttestationInformationVerification(AttestationVerificationResult result)
 {
-    if (err != CHIP_NO_ERROR)
+    StopSessionIf(mState != State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    StopSessionIf(mDeviceBeingCommissioned == nullptr, CHIP_ERROR_INCORRECT_STATE);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1152]: These checks did not use to be in `OnDeviceAttestationInformationVerification`.  Why are they being added?  Is it because we used to cancel the callback when either of those things became true and now we no longer use cancelable callbacks?

File: src/controller/CHIPDeviceController.cpp [hunk 7]
```diff
@@ -1209,192 +1143,167 @@ CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(const ByteSpan & attestat
...
-    {
-        ChipLogError(Controller, "Failed in generating device's operational credentials. Error %s", ErrorStr(err));
-        commissioner->OnSessionEstablishmentError(err);
+        StopSessionIf(ConvertX509CertToChipCert(noc, nocCert));
+        StopSessionIf(device->SetNOCCertBufferSize(nocCert.size()));
     }
 }
 
-CHIP_ERROR DeviceCommissioner::ProcessOpCSR(const ByteSpan & NOCSRElements, const ByteSpan & AttestationSignature)
+CHIP_ERROR DeviceCommissioner::SendTrustedRootCertificate(CommissioneeDeviceProxy * device, const ByteSpan & rcac)
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1261]: This lost the logging it used to have, no?

File: src/controller/CHIPDeviceController.cpp [hunk 8]
```diff
@@ -1209,192 +1143,167 @@ CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(const ByteSpan & attestat
...
-    mOperationalCredentialsDelegate->SetFabricIdForNextNOCRequest(fabric->GetFabricId());
+    OperationalCredentials::Commands::AddTrustedRootCertificate::Type request;
+    request.rootCertificate = rcac;
+    return SendCommand<OperationalCredentialsCluster>(device, request, OnAddTrustedRootCertificate);
+}
 
-    return mOperationalCredentialsDelegate->GenerateNOCChain(NOCSRElements, AttestationSignature, device->GetDAC(), ByteSpan(),
-                                                             ByteSpan(), &mDeviceNOCChainCallback);
+TRAMPOLINE(AddTrustedRootCertificate, const app::DataModel::NullObjectType &);
+void DeviceCommissioner::OnAddTrustedRootCertificate(const app::DataModel::NullObjectType & data)
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1269]: This also lost its logging....

File: src/controller/CHIPDeviceController.cpp [hunk 9]
```diff
@@ -1209,192 +1143,167 @@ CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(const ByteSpan & attestat
...
 
 CHIP_ERROR DeviceCommissioner::SendOperationalCertificate(CommissioneeDeviceProxy * device, const ByteSpan & nocCertBuf,
                                                           const ByteSpan & icaCertBuf)
 {
-    VerifyOrReturnError(device != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
-    chip::Controller::OperationalCredentialsCluster cluster;
-    cluster.Associate(device, 0);
+    OperationalCredentials::Commands::AddNOC::Type request;
+    request.NOCValue      = mDeviceBeingCommissioned->GetNOCCert();
+    request.ICACValue     = MakeOptional(mDeviceBeingCommissioned->GetICACert());
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1282]: This used to send `nocCertBuf` and `icaCertBuf`.  Why the change to not use the arguments?

File: src/controller/CHIPDeviceController.cpp [hunk 10]
```diff
@@ -1209,192 +1143,167 @@ CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(const ByteSpan & attestat
...
 
-    Callback::Cancelable * successCallback = mNOCResponseCallback.Cancel();
-    Callback::Cancelable * failureCallback = mOnCertFailureCallback.Cancel();
+    return SendCommand<OperationalCredentialsCluster>(mDeviceBeingCommissioned, request, OnAddNOCResponse);
+}
 
-    ReturnErrorOnFailure(cluster.AddNOC(successCallback, failureCallback, nocCertBuf, icaCertBuf, ByteSpan(nullptr, 0),
-                                        mLocalId.GetNodeId(), mVendorId));
+TRAMPOLINE(AddNOCResponse, const OperationalCredentials::Commands::NOCResponse::DecodableType &);
+void DeviceCommissioner::OnAddNOCResponse(const OperationalCredentials::Commands::NOCResponse::DecodableType & data)
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1291]: This also lost its logging....  Several levels of it, in fact.

File: src/controller/CHIPDeviceController.cpp [hunk 11]
```diff
@@ -1744,16 +1518,11 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
     switch (step)
     {
     case CommissioningStage::kArmFailsafe: {
-        // TODO(cecille): This is NOT the right way to do this - we should consider attaching an im delegate per command or
-        // something. Per exchange context?
-        ChipLogProgress(Controller, "Arming failsafe");
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1749]: Why is this log being removed?

File: src/controller/CHIPDeviceController.cpp [hunk 12]
```diff
@@ -1765,51 +1534,41 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
         // TODO(cecille): Set time as well once the time cluster is implemented
         // TODO(cecille): Worthwhile to keep this around as part of the class?
         // TODO(cecille): Where is the country config actually set?
-        ChipLogProgress(Controller, "Setting Regulatory Config");
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1768]: Why did this log go away?

File: src/controller/CHIPDeviceController.cpp [hunk 13]
```diff
@@ -1765,51 +1534,41 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
-#else
-        status = CHIP_ERROR_NOT_IMPLEMENTED;
+        hasRegulatoryLocation = (CHIP_NO_ERROR == DeviceLayer::ConfigurationMgr().GetRegulatoryLocation(regulatoryLocation));
+        hasCountryCode =
+            (CHIP_NO_ERROR ==
+             DeviceLayer::ConfigurationMgr().GetCountryCode(countryCodeStr, kMaxCountryCodeSize, actualCountryCodeSize));
 #endif
-        if (status != CHIP_NO_ERROR)
+
+        if (hasRegulatoryLocation)
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1551]:
```suggestion
        if (!hasRegulatoryLocation)
```

File: src/controller/CHIPDeviceController.cpp [hunk 14]
```diff
@@ -1765,51 +1534,41 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
+        {
+            ChipLogError(Controller, "Unable to find regulatory location, defaulting to outdoor");
+        }
+
+        if (hasCountryCode)
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1556]:
```suggestion
        if (!hasCountryCode)
```

File: src/controller/CHIPDeviceController.cpp [hunk 15]
```diff
@@ -1765,51 +1534,41 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
             ChipLogError(Controller, "Unable to find country code, defaulting to WW");
         }
-        chip::CharSpan countryCode(countryCodeStr, actualCountryCodeSize);
 
-        GeneralCommissioningCluster genCom;
-        genCom.Associate(proxy, 0);
-        genCom.SetRegulatoryConfig(mSuccess.Cancel(), mFailure.Cancel(), regulatoryLocation, countryCode, breadcrumb,
-                                   kCommandTimeoutMs);
+        GeneralCommissioning::Commands::SetRegulatoryConfig::Type request;
+        request.location    = static_cast<EmberAfRegulatoryLocationType>(regulatoryLocation);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1562]: This needs to cast to the right type (the cluster-objects one) to compile on tip....  `EmberAfRegulatoryLocationType` is gone.

File: src/controller/CHIPDeviceController.cpp [hunk 16]
```diff
@@ -1765,51 +1534,41 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
+        request.countryCode = CharSpan(countryCodeStr, actualCountryCodeSize);
+        request.breadcrumb  = breadcrumb;
+        request.timeoutMs   = kCommandTimeoutMs;
+        StopSessionIf(SendCommand<GeneralCommissioningCluster>(proxy, request, OnSuccess<decltype(request)::ResponseType>));
     }
     break;
     case CommissioningStage::kDeviceAttestation: {
-        ChipLogProgress(Controller, "Exchanging vendor certificates");
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1802]: Why did this logging go away?

File: src/controller/CHIPDeviceController.cpp [hunk 17]
```diff
@@ -1765,51 +1534,41 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
-        // TODO(cecille): Remove the certificates from the CommissioneeDeviceProxy and take from the commissioning parameters.
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1803]: Why did the TODO comment go away?

File: src/controller/CHIPDeviceController.cpp [hunk 18]
```diff
@@ -1765,51 +1534,41 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
-        CHIP_ERROR status =
-            SendCertificateChainRequestCommand(reinterpret_cast<CommissioneeDeviceProxy *>(proxy), CertificateType::kPAI);
-        if (status != CHIP_NO_ERROR)
-        {
-            ChipLogError(Controller, "Failed in sending 'Certificate Chain Request' command to the device: err %s",
-                         ErrorStr(status));
-            OnSessionEstablishmentError(status);
-            return;
-        }
+        StopSessionIf(
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1570]: And this one should probably be named `StopSessionAndReturnOnFailure` or something...  `StopSessionIf` is really confusing when there is no boolean being passed to it.

File: src/controller/CHIPDeviceController.cpp [hunk 19]
```diff
@@ -1842,11 +1595,12 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
             mCommissioningDelegate->CommissioningStepFinished(CHIP_ERROR_INVALID_ARGUMENT, report);
             return;
         }
-        ChipLogProgress(Controller, "Adding wifi network");
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1845]: Logging got lost.

File: src/controller/CHIPDeviceController.cpp [hunk 20]
```diff
@@ -1857,11 +1611,11 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
             mCommissioningDelegate->CommissioningStepFinished(CHIP_ERROR_INVALID_ARGUMENT, report);
             return;
         }
-        ChipLogProgress(Controller, "Adding thread network");
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1860]: Logging got lost?

File: src/controller/CHIPDeviceController.cpp [hunk 21]
```diff
@@ -1872,10 +1626,11 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
             mCommissioningDelegate->CommissioningStepFinished(CHIP_ERROR_INVALID_ARGUMENT, report);
             return;
         }
-        ChipLogProgress(Controller, "Enabling wifi network");
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1875]: Logging got lost.

File: src/controller/CHIPDeviceController.cpp [hunk 22]
```diff
@@ -1891,21 +1646,20 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
             mCommissioningDelegate->CommissioningStepFinished(CHIP_ERROR_INVALID_ARGUMENT, report);
             return;
         }
-        ChipLogProgress(Controller, "Enabling thread network");
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1894]: Logging got lost.

File: src/controller/CHIPDeviceController.cpp [hunk 23]
```diff
@@ -1891,21 +1646,20 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
...
+        request.breadcrumb = breadcrumb;
+        StopSessionIf(SendCommand<NetworkCommissioningCluster>(proxy, request, OnSuccess<decltype(request)::ResponseType>));
     }
     break;
     case CommissioningStage::kFindOperational: {
         UpdateDevice(proxy->GetDeviceId());
     }
     break;
     case CommissioningStage::kSendComplete: {
-        ChipLogProgress(Controller, "Calling commissioning complete");
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1905]: Logging got lost.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/controller/CHIPDeviceController.h [hunk 24]
```diff
@@ -714,108 +715,84 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
...
+    static void OnSuccess(void * context, const T & data)
+    {
+        DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+        commissioner->CommissioningStageComplete(CHIP_NO_ERROR);
+    }
+
+    static void OnFailure(void * context, EmberAfStatus status)
+    {
+        DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+        commissioner->SessionEstablishmentErrorIf(true, static_cast<CHIP_ERROR>(status));
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.h:771]: This static_cast does not make sense.

I think the old code just used a generic failure code.  We can do that, or we can create a StatusIB from the EmberAfStatus (by converting it to an `InteractionModel::Status`; we have functions for that), and using the new "put a StatusIB into a CHIP_ERROR" bits we have now.

**Comment by woody-apple** [comment]
Fast tracking, given > 3 days for review.

File: src/controller/CHIPDeviceController.cpp (hunk 2 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have removed the macros.

File: src/controller/CHIPDeviceController.h [hunk 25]
```diff
@@ -415,8 +413,11 @@ class DLL_EXPORT DeviceController : public SessionReleaseDelegate,
 
     CommissioningWindowOption mCommissioningWindowOption;
 
-    static void OnOpenPairingWindowSuccessResponse(void * context);
-    static void OnOpenPairingWindowFailureResponse(void * context, uint8_t status);
+    TRAMPOLINE(OpenPairingWindowSuccessResponse, const app::DataModel::NullObjectType &);
```
[the remark this answers is not in the document]
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.h]: I have removed the macros.

File: src/controller/CHIPDeviceController.cpp (hunk 4 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have reverted it back.

File: src/controller/CHIPDeviceController.cpp (hunk 20 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp:1819]: I have reverted it back.

File: src/controller/CHIPDeviceController.cpp (hunk 19 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp:1805]: I have reverted it back.

File: src/controller/CHIPDeviceController.cpp (hunk 21 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp:1833]: I have reverted it back.

File: src/controller/CHIPDeviceController.cpp (hunk 22 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp:1850]: I have reverted it back.

File: src/controller/CHIPDeviceController.cpp (hunk 23 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp:1867]: I have reverted it back.

File: src/controller/CHIPDeviceController.cpp (hunk 3 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have updated the code have as few changes as possible. So `StopSessionIf` is gone.

File: src/controller/CHIPDeviceController.cpp (hunk 5 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have updated the code to look much closer to the previous state. So those are gone.

File: src/controller/CHIPDeviceController.cpp (hunk 7 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have added it back.

File: src/controller/CHIPDeviceController.cpp (hunk 8 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have added it back.

File: src/controller/CHIPDeviceController.cpp (hunk 9 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have reverted that.

File: src/controller/CHIPDeviceController.cpp (hunk 10 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have added it back.

File: src/controller/CHIPDeviceController.cpp (hunk 11 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp:1639]: I have added it back.

File: src/controller/CHIPDeviceController.cpp (hunk 12 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have added it back.

File: src/controller/CHIPDeviceController.cpp (hunk 14 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have reverted all those changes.

File: src/controller/CHIPDeviceController.cpp (hunk 15 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have reverted all those changes.

File: src/controller/CHIPDeviceController.cpp (hunk 16 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have added it back.

File: src/controller/CHIPDeviceController.cpp (hunk 17 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have reverted this changes.

File: src/controller/CHIPDeviceController.cpp (hunk 18 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have removed `StopSessionIf`.

File: src/controller/CHIPDeviceController.cpp (hunk 13 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.cpp]: I have reverted this change.

File: src/controller/CHIPDeviceController.h (hunk 24 above)
Comment (vivien-apple) [reply src/controller/CHIPDeviceController.h]: Hopefully we are using `CHIP_ERROR` after your changes and the conversion to `Invoke` !

**Comment by vivien-apple** [comment]
To minimise the amount of changes I have rewritten this PR to "only" contains conversion from `CHIPCallback` to the `InvokeCommand` API.

**Review by bzbarsky-apple — APPROVED** [review]
